WalletWasabi/WabiSabi/Models/Decomposition/PossibleDecompositions.cs

```diff
@@ -0,0 +1,194 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using NBitcoin;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.WabiSabi.Models.Decomposition
+{
+	public class PossibleDecompositions
+	{
+		public PossibleDecompositions(
+			IEnumerable<Money> nominalValues,
+			Money maximumTotalValue,
+			Money minimumTotalValue,
+			int maxOutputs)
+		{
+			// 8 is more than enough to be unreasonable in practically any
+			// situation.
+			Guard.InRangeAndNotNull(nameof(maxOutputs), maxOutputs, 1, 8);
+
+			var orderedDenoms = nominalValues.Where(x => x <= maximumTotalValue).OrderByDescending(x => x).ToImmutableArray();
+
+			MaximumTotalValue = maximumTotalValue;
+			MinimumTotalValue = minimumTotalValue;
+
+			// TODO support different effective cost prune ranges for different
+			// sized combinations, using a higher max value and 0 minimum for up
+			// to size 4 to query likely denoms, but only the wallet's
+			// registered balance and tightly bounded objective loss for
+			// optimizing decomposition, to improve efficiency.
+			if (maxOutputs == 1)
+			{
+				var prunedSingletons = new CombinationsOfASize(orderedDenoms, maximumTotalValue, minimumTotalValue);
+				StratifiedDecompositions = ImmutableArray.Create<CombinationsOfASize>(prunedSingletons);
+			}
+			else
+			{
+				var bySize = ImmutableArray.CreateBuilder<CombinationsOfASize>(maxOutputs);
+
+				// Generate the base decompositions, one for each possible value,
+				// without pruning by minimum total value.
+				var unprunedSingletons = new CombinationsOfASize(orderedDenoms, maximumTotalValue, 0);
```

Comment:
Not sure if this should be fixed in wasabi or nbitcoin:

![image](https://user-images.githubusercontent.com/13405205/141213338-f9747b6c-5eae-477f-aed9-1ee136ba1b85.png)

Reply:
NBitcoin is correct, I will fix it